packages/compass-import-export/src/export/gather-fields.ts

```diff
@@ -138,32 +95,185 @@ export function createProjectionFromSchemaFields(fields: SchemaPath[]) {
+      schemaAnalyzer.analyzeDoc(doc);
+      result.docsProcessed++;
+      progressCallback?.(result.docsProcessed);
+      callback();
+    },
+  });
+
+  try {
+    await pipeline(
+      [input, analyzeStream],
+      ...(abortSignal ? [{ signal: abortSignal }] : [])
+    );
+  } catch (err: any) {
+    if (err.code === 'ABORT_ERR') {
+      result.aborted = true;
+    } else {
+      throw err;
+    }
+  }
+
+  // TODO: finalizeSchema() inside schema analyzer replaces the fields object
+  // internally with something of a different type. We should fix that by making
+  // a different final result object that matches the type we want. COMPASS-6720
+  const fields = (
+    (schemaAnalyzer.getResult() as any).fields as BSONSchemaField[]
+  ).filter((field) => !isInternalFieldPath(field.path));
+
+  return {
+    paths: schemaToPaths(fields),
+    ...result,
+  };
+}
+
+// At the time of writing the export UI won't use this code with an aggregation,
+// so this is just included for completeness. ie. we can determine the list of
+// fields for any find query or aggregation.
+export async function gatherFieldsFromAggregation({
+  ns,
+  dataService,
+  aggregation,
   sampleSize,
-}: GatherFieldsOptions): Promise<SchemaPath[]> {
-  const schema = await analyzeSchema({
-    dataService,
-    abortSignal,
+  ...exportOptions
+}: Omit<GatherFieldsOptions, 'input'> & {
+  ns: string;
+  dataService: DataService;
+  aggregation: ExportAggregation;
+  sampleSize?: number;
+}) {
+  debug('gatherFieldsFromAggregation()', { ns: toNS(ns) });
+
+  const { stages, options: aggregationOptions = {} } = aggregation;
+  aggregationOptions.maxTimeMS = capMaxTimeMSAtPreferenceLimit(
+    aggregationOptions.maxTimeMS
+  );
+
+  // optionally sample
+  const sampleStages = stages.slice();
+  if (sampleSize && sampleSize > 0) {
+    sampleStages.push({
+      $sample: {
+        size: sampleSize,
+      },
+    });
+  }
+
+  const aggregationCursor = dataService.aggregateCursor(
     ns,
-    filter,
-    sampleSize,
+    sampleStages,
+    aggregationOptions
+  );
+
+  const input = aggregationCursor.stream();
+
+  try {
+    return await _gatherFields({
+      input,
+      ...exportOptions,
+    });
+  } finally {
+    void aggregationCursor.close();
+  }
+}
+
+function capLimitToSampleSize(limit?: number, sampleSize?: number) {
+  if (limit) {
+    if (sampleSize) {
+      return Math.min(limit, sampleSize);
+    } else {
+      return limit;
+    }
+  } else {
+    if (sampleSize) {
+      return sampleSize;
+    } else {
+      return undefined;
+    }
+  }
+}
+
+export async function gatherFieldsFromQuery({
+  ns,
+  dataService,
+  query = { filter: {} },
+  sampleSize,
+  ...exportOptions
+}: Omit<GatherFieldsOptions, 'input'> & {
+  ns: string;
+  dataService: DataService;
+  query?: ExportQuery;
+  sampleSize?: number;
+}) {
+  debug('gatherFieldsFromQuery()', { ns: toNS(ns) });
+
+  const findCursor = dataService.findCursor(ns, query.filter ?? {}, {
```

Comment:
Notice that find doesn't attempt to sample. This is what we do in the old export and after some discussion with @mcasimir we decided that this is good enough here as well. Making aggregations using $sample for what's likely a find query would just open a whole new can of worms.